semantic.js - fix function arg count

bkutil opened the pull request:
For a function with single required argument, the total number of
arguments would be computed as NaN.

This was caused by the value of the field holding optional number
of arguments being 'undefined' and the addition with an integer
converting the result into NaN.

Fix: #328

bkutil commented:
cc: @dmajda

dmajda commented:
I think this should be tested in `test/runtime/specs/juttle-spec/expressions/function-call.spec.md`. I’d just:

  1. Clone the parameters testcases.
  2. Remove optional parameters in one copy and adapt (e.g. remove now irrelevant testcases).
  3. Add `(with optional parameters)` and `(without optional parameters)` suffixes to the descriptions.

dmajda commented:
The commit message could use some improvement. Without looking at the diff I wasn’t able to get what the change is about. The message mainly needs more context (that the computation is in semantic pass which produces the faulty message described in the bug).

bkutil commented:
Updated the commit message, simplified the code and added tests. Also caught a small bug in passing array argument to the error message for interpolation instead of a string.

dmajda commented on test/runtime/specs/juttle-spec/expressions/function-call.spec.md:
```diff
@@ -76,8 +76,8 @@ Produces an error when passed too many parameters
 
-Returns correct result when passed required parameters (module)
----------------------------------------------------------------
+Returns correct result when passed required parameters (module) (with optional parameters)
```
I’d just do `(module, with optional parameters)`. We do it like this elsewhere IIRC.

dmajda commented on test/runtime/specs/juttle-spec/expressions/function-call.spec.md:
```diff
@@ -151,3 +151,106 @@ Produces an error when passed too many parameters (module)
+
+### Juttle
+
+    function f(a, b) {
```
I’d just do `f(a)` in these tests, this would correspond to the existing tests a little bit more.

(We should really have more proper tests, passing 0, 1, and many parameters, but this is something for another day and another PR.)

dmajda commented on test/runtime/specs/juttle-spec/expressions/function-call.spec.md:
```diff
@@ -151,3 +151,106 @@ Produces an error when passed too many parameters (module)
   * Error: function m.f expects 1 to 2 arguments but was called with 3
+
+Returns correct result when passed required parameters (without optional parameters)
+---------------------------------------------------------------------------------
```
Make `-`’s match the length of the title?

(We should probably switch to `##` for JuttleSpec headers globally at some point, managing the `-`’s is pain.)

dmajda commented:
+1 with the suggested changes

bkutil commented:
Added/removed the missing underlines and fixed parens in test case titles, functions in new tests accept and return only 1 argument, which required fixing also test messages.

I'll merge on green.
